storage/shard.go

```diff
@@ -1542,6 +1571,68 @@ func (s *dbShard) Flush(
 	return s.markFlushStateSuccessOrError(blockStart, multiErr.FinalError())
 }
 
+func (s *dbShard) Snapshot(
+	blockStart time.Time,
+	snapshotStart time.Time,
+	flush persist.Flush,
+) error {
+	// We don't snapshot data when the shard is still bootstrapping
+	s.RLock()
+	if s.bs != bootstrapped {
+		s.RUnlock()
+		return errShardNotBootstrappedToSnapshot
+	}
+	s.RUnlock()
+
+	var multiErr xerrors.MultiError
+
+	s.markIsSnapshotting()
+	defer func() {
+		s.markDoneSnapshotting(multiErr.FinalError() == nil, snapshotStart)
+	}()
+
+	prepareOpts := persist.PrepareOptions{
+		NsMetadata: s.namespace,
+		Shard:      s.ID(),
+		BlockStart: blockStart,
+		WrittenAt:  snapshotStart,
+		IsSnapshot: true,
+	}
+	prepared, err := flush.Prepare(prepareOpts)
+	multiErr = multiErr.Add(err)
+
+	// No action is necessary therefore we bail out early and there is no need to close.
+	if prepared.Persist == nil {
+		// This should never happen in practice, but if it does it means the snapshot file already
+		// exists.
+		s.logger.WithFields(
+			xlog.NewField("blockStart", blockStart),
+			xlog.NewField("snapshotStart", snapshotStart),
+			xlog.NewField("shard", s.ID()),
+		).Errorf("Tried to write prepare snapshot file that already exists")
+		return nil
+	}
+
+	tmpCtx := context.NewContext()
+	s.forEachShardEntry(func(entry *dbShardEntry) bool {
+		series := entry.series
+		// Use a temporary context here so the stream readers can be returned to
```

Comment:
would it help perf if we amortized releasing in batch - i.e. use `s.forEachShardEntryBatch(...)` instead. and only do the BlockingClose after the entire batch?

Reply:
hmm I'm not really sure. I don't think so? The blocking close is just going to put a few items back in the pool and such and I can't easily convert those to bulk operations so I don't think overall throughput would be improed

Reply:
Yeah I'm not certain, but might be worth trying the following version to see if there's any perf difference? 

```
s.forEachShardEntryBatch(func(currEntries []*dbShardEntry) bool {
  tmpCtx.Reset()
  for _, entry := range currEntries {
    multiErr = multiErr.Add(series.Snapshot(tmpCtx, blockStart, prepared.Perist))
  }
  tmpCtx.BlockingClose()
  return true
})

Reply:
Made an issue for it: #516

There is a lot going on in this P.R right now, want to focus on landing a correct implementation and I'll do more benchmarking later (planning on doing a lot of benchmarking of the write/read path once I have functioning implementations that dont lose data)

Reply:
That's fair.